### PR TITLE
fix(sockets): correctly handle pending writes to prevent double write

### PR DESCRIPTION
**Pull request by Threated**
The previous implementation called `send` on the sink directly without using the `SinkWriter` adapter to convert it to an `AsyncWrite`. This meant polling the returned future ourself which seemed to work fine for the majority of cases as it either returned `Poll::Ready` when we were able to send the item to the underlying io or `Poll::Pending` when the io was busy matching the expected behavior of the `AsyncWrite` trait. However there is one other case where the future returned by `send` might return `Poll::Pending` which is when it was able to send the data to the io but flushing the underlying io returned `Poll::Pending`. This is an unfornuate property of the send function which I did not realize when initially implementing it. In pratice this leads to a lot of flushes and in my tests caused the bug that occasionally a packet was sent twice as it has been written to the io but not flushed meansing the future returned `Poll::Pending` leading to the buffer being sent a second time.

TLDR:
The new implementation uses the `SinkWriter` which both does not flush on every write and ensures data is written properly.

### CODE REVIEW

**Review by TKussel — APPROVED** [review]

File: proxy/src/serve_sockets.rs [hunk 1]
```diff
@@ -466,20 +464,35 @@ mod tests {
     async fn test_encryption() {
         let mut key = GenericArray::default();
         OsRng.fill_bytes(&mut key);
-        const N: usize = 2_usize.pow(13);
-        let test_data: &mut [u8; N] = &mut [0; N];
-        OsRng.fill_bytes(test_data);
-        let mut read_buf = [0; N];
+        let data: Arc<Vec<_>> = (0..13337).map(|_| {
+            let mut chunk = vec![0; OsRng.gen_range(1..9999)];
```
[the remark this answers is not in the document]
Comment (Threated) [reply proxy/src/serve_sockets.rs:468]: 9999 because its larger than 8192 which is the default buffer size of `tokio::io::copy_bidirectional` and 13337 because its a cool number and approximately the number of writes that SEL did in my test case.
